src/viam/sdk/common/proto_convert.hpp

```diff
@@ -0,0 +1,90 @@
+#pragma once
+
+#include <boost/callable_traits/args.hpp>
+#include <boost/callable_traits/return_type.hpp>
+#include <boost/mp11/algorithm.hpp>
+
+#include <type_traits>
+
+#define VIAM_SDK_API_FWD_NAMESPACE_BEGIN(module_ns_name) \
+    namespace viam {                                     \
+    namespace module_ns_name {                           \
+    namespace v1 {
+
+#define VIAM_SDK_API_FWD_NAMESPACE_END \
+    }                                  \
+    }                                  \
+    }
+
+namespace viam {
+namespace sdk {
+
+namespace proto_convert_details {
+
+// This struct should be explicitly specialized with a
+//      void operator()(const SdkType&, common::v1::ApiType*) const
+// to provide API/ABI insulated proto conversion
+template <typename SdkType>
+struct to_proto;
+
+// This struct should be explicitly specialized with a
+//      SdkType operator()(const ProtoType*) const
+// to provided API/ABI insulated construction from proto
+template <typename ProtoType>
+struct from_proto;
+
+template <typename Callable>
```

Comment:
- I think this type is worth a comment, especially with the tricky `args_t` thing happening.
- When we get to C++[17/20] does the standard give us tools to do this rather than needing boost? Mostly, is there a path to a polyfill here?

Reply:
It honestly might even be doable in the current standard, just a little horrible maybe. I like callable traits especially for ease of use though especially given the following caveat: https://www.boost.org/doc/libs/1_86_0/libs/callable_traits/doc/html/index.html#callable_traits.introduction.motivation

Reply:
That's pretty horrific. Maybe throw a TODO on here that it would be nice to eliminate this boost dependency in our headers if possible? Maybe future us finds a nice way to do it in some future standard.